### ISSUE TITLE
`Crypto::Subtle.constant_time_compare` does not work with `StaticArray` / Cannot overwrite `StaticArray` variable

### ISSUE THREAD

**Issue by BlobCodes** [crystal-lang/crystal#14610]
## Bug Report

Trying to compare a `StaticArray` with another `StaticArray` with the same size and underlying type always returns true while comparing a `StaticArray` with something else seems to be undefined:
```cr
array1 = uninitialized StaticArray(UInt8, 3)
    (0...3).each { |i| array1[i] = 2_u8 }

array2 = Slice(UInt8).literal(2_u8, 2_u8, 2_u8)

puts array1.to_slice # => Bytes[2, 2, 2]
puts array2.to_slice # => Bytes[2, 2, 2]
puts Crypto::Subtle.constant_time_compare(array1, array2) # => false
```

The underlying issue seems to be that it's not possible to overwrite a staticarray variable with something else but still use it:
```cr
def foo(x)
  x = x.to_slice
  puts x[0]
end

arr = StaticArray(UInt8, 10).new { 1_u8 }
foo(arr) # => 10

arr = StaticArray(UInt8, 29).new { 1_u8 }
foo(arr) # => 29
```

**Comment by straight-shoota** [comment]
Yes, problem lies in the assignment `x = x.to_slice`. `StaticArray#to_slice` takes a pointer of itself. So `x.to_slice` points to the location of `x` on the stack. By assigning the resulting slice to `x` again, you're overriding that memory with the instance of `Slice` which now points to itself.

Reduced example:
```cr
ary = StaticArray[1, 2, 3, 4]
slice = ary.to_slice

slice # => Slice[1, 2, 3, 4]

# now overwrite the memory location of `ary` with a new value
ary = StaticArray[5, 6, 7, 8]

slice # => Slice[5, 6, 7, 8]
```

This is obviously a foot gun.
`StaticArray#to_slice` is unsafe because it takes a stack pointer. It should be renamed to `#to_unsafe_slice`. This has been discussed previously in #9606 and there's already a patch in #11031. I'll extract the part for `StaticArray#to_unsafe_slice`  from that PR which seems to be uncontested.
